x/tokenfactory/keeper/msg_server_test.go

```diff
@@ -0,0 +1,249 @@
+package keeper_test
+
+import (
+	"fmt"
+
+	"github.com/osmosis-labs/osmosis/v11/x/tokenfactory/keeper"
+	"github.com/osmosis-labs/osmosis/v11/x/tokenfactory/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+// TestMintDenomMsg tests TypeMsgMint message is emitted on a successful mint
+func (suite *KeeperTestSuite) TestMintDenomMsg() {
+	// Create a denom
+	suite.CreateDefaultDenom()
+
+	for _, tc := range []struct {
+		desc                  string
+		amount                int64
+		mintDenom             string
+		admin                 string
+		valid                 bool
+		expectedMessageEvents int
+	}{
+		{
+			desc:      "denom does not exist",
+			amount:    10,
+			mintDenom: "factory/osmo1t7egva48prqmzl59x5ngv4zx0dtrwewc9m7z44/evmos",
+			admin:     suite.TestAccs[0].String(),
+			valid:     false,
+		},
+		{
+			desc:                  "success case",
+			amount:                10,
+			mintDenom:             suite.defaultDenom,
+			admin:                 suite.TestAccs[0].String(),
+			valid:                 true,
+			expectedMessageEvents: 1,
+		},
+	} {
+		suite.Run(fmt.Sprintf("Case %s", tc.desc), func() {
+			ctx := suite.Ctx.WithEventManager(sdk.NewEventManager())
+			suite.Require().Equal(0, len(ctx.EventManager().Events()))
+			// Test mint message
+			suite.msgServer.Mint(sdk.WrapSDKContext(ctx), types.NewMsgMint(tc.admin, sdk.NewInt64Coin(tc.mintDenom, 10)))
+			// Ensure current number and type of event is emitted
+			suite.AssertEventEmitted(ctx, types.TypeMsgMint, tc.expectedMessageEvents)
+		})
+	}
+}
+
+// TestBurnDenomMsg tests TypeMsgBurn message is emitted on a successful burn
+func (suite *KeeperTestSuite) TestBurnDenomMsg() {
+	var addr0bal int64
+	// Create a denom.
+	suite.CreateDefaultDenom()
+	// mint 10 default token for testAcc[0]
+	suite.msgServer.Mint(sdk.WrapSDKContext(suite.Ctx), types.NewMsgMint(suite.TestAccs[0].String(), sdk.NewInt64Coin(suite.defaultDenom, 10)))
+	addr0bal += 10
+
+	for _, tc := range []struct {
+		desc                  string
+		amount                int64
+		burnDenom             string
+		admin                 string
+		valid                 bool
+		expectedMessageEvents int
+	}{
+		{
+			desc:      "denom does not exist",
+			burnDenom: "factory/osmo1t7egva48prqmzl59x5ngv4zx0dtrwewc9m7z44/evmos",
+			admin:     suite.TestAccs[0].String(),
+			valid:     false,
+		},
+		{
+			desc:                  "success case",
+			burnDenom:             suite.defaultDenom,
+			admin:                 suite.TestAccs[0].String(),
+			valid:                 true,
+			expectedMessageEvents: 2, // TODO: why is this two
```

Comment:
Is it possible that these are emitted in both the message server and in the keeper?

I found that to be common in other modules

Reply:
I was searching where that specific type of event is emitted and found that it was erroneously duplicated on the execution path. 

The quickest way to verify that those paths were emitting the events was to set breakpoints in all places where that event is emitted and confirm that they were being hit in different locations.
 
I haven't checked but suspect that something similar might be here too.

Reply:
Hmm so it turns out the bankKeeper is emitting a MsgBurn here https://github.com/osmosis-labs/osmosis/blob/83bd44e30067d5c5a147361696a413f45134b708/x/tokenfactory/keeper/bankactions.go#L51-L52

As well as we emit an event here https://github.com/osmosis-labs/osmosis/blob/83bd44e30067d5c5a147361696a413f45134b708/x/tokenfactory/keeper/msg_server.go#L95-L101

Unsure if I should just remove the manually emitted event or just accept two events being emitted for a single event. Wdyt?

Reply:
Ahh I see so it seems that the event names for two different components are the same. I think we might consider renaming the tokenfactory burn event since it clashes with bank/

Reply:
It might make sense to rename most of these by prefixing them with something else such as `tf_mint`, `tf_burn`
https://github.com/osmosis-labs/osmosis/blob/eacd4e21d4b994fb67e7087aa2134d7bdc397d5c/x/tokenfactory/types/msgs.go#L11-L16

The reason is that many of these event names clash with other sdk events so they can't really be useful IMO

Reply:
If we go with this approach, we need to make sure to have a changelog entry

Reply:
ahhh I wasn't sure where the name clash was happening between sdk and here but that makes sense! Agreed on changing with the tf prefix, will do that and add changelog, thanks!

Reply:
I double-checked and, actually, "burn" is the only clash.

I still think that we should be prefixing events by module names to avoid such issues. However, we can create a separate issue to discuss this.

For now, I think the approach would be to only rename the tokenfactory "burn" event to mitigate the immediate problem.

Please let me know what you think

Reply:
Makes sense to me, thank you